src/Common/GuardedPoolAllocator.cpp

```diff
@@ -0,0 +1,263 @@
+        "Failed to allocate in guarded pool allocator memory");
+}
+
+void GuardedPoolAllocator::deallocateInGuardedPool(void *ptr,
+                                                   size_t size) const {
+    assert((reinterpret_cast<uintptr_t>(ptr) % state.pageSize) == 0);
+    assert((size % state.pageSize) == 0);
+    // mmap() a PROT_NONE page over the address to release it to the system, if
+    // we used mprotect() here the system would count pages in the quarantine
+    // against the RSS.
+    check(mmap(ptr, size, PROT_NONE, MAP_FIXED | MAP_ANONYMOUS | MAP_PRIVATE, -1,
+                0) != MAP_FAILED,
+        "Failed to deallocate in guarded pool allocator memory");
+}
+
+uintptr_t GuardedPoolAllocator::alignUp(uintptr_t ptr, size_t alignment) {
+    assert(isPowerOfTwo(alignment));
+    assert(alignment != 0);
+    if ((ptr & (alignment - 1)) == 0)
+        return ptr;
+
+    printf("Align up; add %lu to pointer\n", alignment - (ptr & (alignment - 1)));
+    printf("Old=%p; new=%p\n", reinterpret_cast<void *>(ptr), reinterpret_cast<void *>(ptr + alignment - (ptr & (alignment - 1))));
+    ptr += alignment - (ptr & (alignment - 1));
+    return ptr;
+}
+
+uintptr_t GuardedPoolAllocator::alignDown(uintptr_t ptr, size_t alignment) {
+    assert(isPowerOfTwo(alignment));
+    assert(alignment != 0);
+    if ((ptr & (alignment - 1)) == 0)
+        return ptr;
+
+    printf("Old=%p; new=%p\n", reinterpret_cast<void *>(ptr), reinterpret_cast<void *>(ptr - (ptr & (alignment - 1))));
+    printf("Align down; subtract %lu from pointer\n", ptr & (alignment - 1));
+    ptr -= ptr & (alignment - 1);
+    return ptr;
+}
+
+void GuardedPoolAllocator::init()
+{
+    printf("Initialize gwp asan\n");
+    singletonPtr = this;
+
+    StackTrace trace;
+    printf("Trace: %s", trace.toString().c_str());
+
+    state.maxSimultaneousAllocations = 4;  // TODO: this must be passed through init options
+
+    const size_t pageSize = getPlatformPageSize();
+    state.pageSize = pageSize;
+
+    printf("Max alloc size: %zu\n", pageSize);
+
+    /// check that page size is a power of 2
+    assert((pageSize & (pageSize - 1)) == 0);
+    size_t poolSize =
+        pageSize * (1 + state.maxSimultaneousAllocations) +
+        state.maxSimultaneousAllocations * state.maximumAllocationSize();
+
+    assert(poolSize % pageSize == 0);
+    void* guardedPoolMemory = reserveGuardedPool(poolSize);
+    state.guardedPagePool = reinterpret_cast<uintptr_t>(guardedPoolMemory);
+    state.guardedPagePoolEnd =
+        reinterpret_cast<uintptr_t>(guardedPoolMemory) + poolSize;
+
+    printf("Guarded pool is at %p\n", reinterpret_cast<void *>(state.guardedPagePool));
+    printf("Guarded pool end is at %p\n", reinterpret_cast<void *>(state.guardedPagePoolEnd));
+
+    size_t freeSlotsRequiredBytes = roundUpTo(
+        state.maxSimultaneousAllocations * sizeof(*freeSlots), pageSize);
+    freeSlots =
+        reinterpret_cast<size_t *>(map(freeSlotsRequiredBytes));
```

Comment:
I guess there is no need in adding `map` method, since it is used only here, all other places that do `mmap` does not uses it, so I would prefer to inline `map` method here instead.

Reply:
In incoming update there will be another use of this function? (for metadata allocation)
Do you think it will still be preferable to inline it?

Reply:
>In incoming update there will be another use of this function

Ok, but it is better to make this function static only (no need to make it a class method)